### PR TITLE
Upgrades the Modsuit Adapter Shell

### PR DESCRIPTION
**Pull request by RikuTheKiller**
## About The Pull Request

Code improvements are much appreciated as some things may be rather hacky.

Adds more options to the currently very limited modsuit adapter shell. Right now you can only select a module and activate (not deploy) the suit.

This has some major problems as you literally can't even deploy the suit to activate it so that's rendered useless and selecting a module is like... kind of a weird input anyways but I won't judge so I left it in. Please comment down below if you'd like for me to add an "Activate Selected Module" input and "On Module Activated" output as those are certainly possible to do. I was just a little torn on how balanced that would be.

Changes:
- "Module to Select" input is now an option. You can still use a string input, but simply inserting it into the suit and activating it, then accessing the circuit that way will give you a list of all modules that the modsuit has.
- Modsuit quick deploy (RMB) no longer tries to deploy the rest of the pieces when used while the suit is only partially deployed. It will now instead retract the extended pieces. This makes the "Toggle Deployment" input less prone to errors. (Why was it like this in the first place? Having to manually retract the already extended pieces sucks ass.)

Added Inputs:
- "Toggle Deployment" is a new signal input that does exactly what it says it does. It simply tries to extend or retract all pieces of the modsuit depending on it's current state.

Added Outputs:
- "Activated" is a new number output that outputs 1 if the suit is activated and 0 if it's not.
- "Deployed" is a new number output that outputs 1 if all parts of the suit are extended and 0 if they aren't.
- "Deployed Parts" is a new string list output that outputs a list of the names of all currently deployed parts.
- "On Deploy" is a new signal output that outputs a signal whenever all parts of the suit are deployed or retracted, regardless of the method used.
- "Finished Toggling" is a new signal output that outputs a signal whenever the suit has finished activating or deactivating, regardless of the method used.

## Why It's Good For The Game

Expanding a shell that's almost completely useless does seem pretty good for the game, at least in my opinion.

It also opens up a bunch of creative options regarding the shell. I plan on expanding the functionality of MMI components and making modsuit AIs able to interact with circuit adapters. (We'll see.)

## Changelog

:cl:
add: Massively expanded the functionality of the modsuit adapter shell
qol: Modsuit quick deploy now retracts the already extended pieces if used while the suit is partially extended
/:cl:

### CODE REVIEW

**Comment by RikuTheKiller** [comment]
Second "oops me dumb" moment, didn't add COMSIG_MOD_MODULE_CHANGED to the removal of a module. Gonna fix that once I get home, do not merge this until then.

File: code/modules/mod/mod_control.dm
Comment (Watermelon914) [inline code/modules/mod/mod_control.dm:365]: use different signals, don't send the same signal in different places. Use signals like COMSIG_MOD_MODULE_REMOVED, COMSIG_MOD_MODULE_ADDED

File: code/modules/wiremod/shell/module.dm
Comment (Watermelon914) [inline code/modules/wiremod/shell/module.dm:140]: please make sure attached_module is valid before doing this.

File: code/modules/wiremod/shell/module.dm
Comment (Watermelon914) [inline code/modules/wiremod/shell/module.dm:132]: in this proc, `RegisterSignal(attached_module, COMSIG_MOVABLE_MOVED, .proc/on_move)` needs to be in the if statement block and not outside of it.

File: code/modules/wiremod/shell/module.dm
Comment (Watermelon914) [inline code/modules/wiremod/shell/module.dm:82]: subtype this as an option port so that you don't have to do any typecasting.

File: code/modules/wiremod/shell/module.dm
Comment (Watermelon914) [inline code/modules/wiremod/shell/module.dm:189]: remove this typecasting, see above comment

File: code/modules/wiremod/shell/module.dm
Comment (Watermelon914) [inline code/modules/wiremod/shell/module.dm:168]: remove this typecasting, see above comment

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/modules/wiremod/shell/module.dm
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/module.dm:132]: I'm probably not even the one who fucked that one up, but yeah.

File: code/modules/wiremod/shell/module.dm
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/module.dm:140]: that was a leftover mistake from some tomfoolery, fixed

File: code/modules/mod/mod_control.dm
Comment (RikuTheKiller) [reply code/modules/mod/mod_control.dm:365]: will do

File: code/modules/wiremod/shell/module.dm
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/module.dm]: done

File: code/modules/wiremod/shell/module.dm
Comment (Watermelon914) [reply code/modules/wiremod/shell/module.dm:132]: yeah, but it's a bug nonetheless that I noticed

File: code/modules/mod/mod_activation.dm
Comment (Watermelon914) [inline code/modules/mod/mod_activation.dm:64]: same thing as before here, use only 1 signal type per event. Put these into a proc instead that gets called, e.g. on_mod_part_toggled

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/modules/wiremod/shell/module.dm
Comment (Watermelon914) [inline code/modules/wiremod/shell/module.dm:203]: honestly, it may just be a good idea to separate these out into two signals, COMSIG_MOD_PART_DEPLOYED and COMSIG_MOD_PART_UNDEPLOYED

Comment (RikuTheKiller) [reply code/modules/wiremod/shell/module.dm:203]: Should I seperate the output signals too on the component?

And should I seperate COMSIG_MOD_TOGGLED too?

File: code/modules/mod/mod_activation.dm
Comment (RikuTheKiller) [reply code/modules/mod/mod_activation.dm:64]: Uh, what do you mean by this? Do the same thing I did with COMSIG_MOD_MODULE_CHANGED?

**Pull request merged**
